### PR TITLE
Fetch minimum JDK version from Jenkinsfile

### PR DESCRIPTION
**Pull request by sridamul**
Fetches the minimum JDK version from Jenkinsfile.
For eg: 
Login-theme-plugin Jenkinsfile
```
buildPlugin(
  useContainerAgent: true, // Set to `false` if you need to use Docker for containerized tests
  configurations: [
    [platform: 'linux', jdk: 21],
    [platform: 'windows', jdk: 17],
])
```
Fetches the min jdk (platform windows) which is `17`

Screenshot 2024-08-20 135232

Now this JDK version can be used directly to compile and verify plugins if the Jenkinsfile is present.

### Testing done
Interactive testing + Unit tests

### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/extractor/PluginMetadata.java [hunk 1]
```diff
@@ -37,6 +37,11 @@ public class PluginMetadata extends CacheEntry<PluginMetadata> implements Serial
      */
     private boolean hasJenkinsfile;
 
+    /**
+     * Minimum JDK version required by the plugin
+     */
+    private int jdkVersion;
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/extractor/PluginMetadata.java:43]: Why an int when we have now an Enum for the JDK model ?
Why not getting list of JDK version configured at let the call (PluginModernizer) use they JDK they need (minimum or maximum)

File: plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollectorTest.java [hunk 2]
```diff
@@ -143,5 +143,7 @@ void testPlugin() throws Exception {
         assertEquals(10, properties.size());
         boolean hasJenkinsfile = pluginMetadata.hasJenkinsfile();
         assertFalse(hasJenkinsfile);
+        int jdkVersion = pluginMetadata.getJdkVersion();
+        assertEquals(0, jdkVersion);
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollectorTest.java:147]: Why not adding a file to the rewrite test or even writting new test ?

```
groovy(
        """
          buildPlugin()
        """,
        spec -> spec.path("Jenkinsfile")),
```

**Review by jonesbusy — CHANGES_REQUESTED** [review]
Thanks for your PR. But there are few things wrong here that need to be fixed.

Also the code need to be maintainable and more easy to read

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollector.java [hunk 3]
```diff
@@ -49,6 +58,7 @@ public static class MetadataAccumulator {
         private final List<ArchetypeCommonFile> commonFiles = new ArrayList<>();
         private final List<String> otherFiles = new ArrayList<>();
         private final List<MetadataFlag> flags = new LinkedList<>();
+        private JDK jdkVersion;
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollector.java:61]: I think we should accumulate all JDK version found and not only the latest

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollector.java [hunk 4]
```diff
@@ -98,8 +108,54 @@ public Tree visit(Tree tree, ExecutionContext ctx) {
...
+                                                            if (mapExpr instanceof G.MapEntry mapEntry) {
+
+                                                                J.Literal key = (J.Literal) mapEntry.getKey();
+                                                                J.Literal value = (J.Literal) mapEntry.getValue();
+
+                                                                if ("jdk".equals(key.getValue())) {
+                                                                    int currentJdkVersion =
+                                                                            Integer.parseInt(value.getValue()
+                                                                                    .toString());
+                                                                    if (currentJdkVersion < jdkVersion) {
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollector.java:142]: This logic should not be there. It's the responsibility of the user of metadata to select the JDK version. The collector should be dumb as possible

File: plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollectorTest.java [hunk 5]
```diff
@@ -164,4 +164,226 @@ void testPlugin() throws Exception {
...
+                         buildPlugin(
+                         useContainerAgent: true,
+                         configurations: [
+                                [platform: 'linux', jdk: 21],
+                                [platform: 'windows', jdk: 21],
+                         ])
+                         """,
+                        spec -> spec.path("Jenkinsfile")),
+                // language=xml
+                pomXml(
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollectorTest.java:183]: This test only check Jenkinsfile. Let's try to use a minimal pom. And if possible from a constant. Will make the test more easy to read

File: plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollectorTest.java [hunk 6]
```diff
@@ -164,4 +164,226 @@ void testPlugin() throws Exception {
...
+                              <id>repo.jenkins-ci.org</id>
+                              <url>https://repo.jenkins-ci.org/public/</url>
+                            </pluginRepository>
+                          </pluginRepositories>
+                        </project>
+                        """));
+        PluginMetadata pluginMetadata = new PluginMetadata().refresh();
+        JDK jdkVersion = pluginMetadata.getJdkVersion();
+        assertEquals(JDK.JAVA_17, jdkVersion);
+    }
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/extractor/MetadataCollectorTest.java:388]: Missing some test with empty buildPlugin() or more outdated/exotic configuration.

For example how the visitors are reacting to https://github.com/jenkinsci/forensics-api-plugin/blob/main/Jenkinsfile

**Review by jonesbusy — CHANGES_REQUESTED** [review]

**Comment by sridamul** [comment]
Hey @jonesbusy, Just now seen your reviews. Let me change accordingly. Thanks!

**Comment by sridamul** [comment]
> Missing some test with empty buildPlugin() or more outdated/exotic configuration.
> 
> For example how the visitors are reacting to https://github.com/jenkinsci/forensics-api-plugin/blob/main/Jenkinsfile

Working on to update the logic, will raise new pr to cover this kind of file too.

**Review by jonesbusy — APPROVED** [review]

**Comment by jonesbusy** [comment]
Thanks a lot! Merging, very happy to start testing this :)
